In [1]:
import os

from langchain.llms import OpenAI
from langchain.prompts import (PromptTemplate, FewShotPromptTemplate)
from langchain.memory import ConversationBufferMemory 
from langchain.chains import (LLMChain, SimpleSequentialChain, SequentialChain)
from langchain.embeddings.openai import OpenAIEmbeddings
from langchain.text_splitter import CharacterTextSplitter
from langchain.vectorstores import FAISS
from langchain.chains.question_answering import load_qa_chain
from typing_extensions import Concatenate
import streamlit as st
import pandas as pd
from PyPDF2 import PdfReader


from transformers import AutoTokenizer, AutoModel


/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:



input_texts = [
    "what is the capital of China?",
    "how to implement quick sort in python?",
    "Beijing",
    "sorting algorithms"
]

tokenizer = AutoTokenizer.from_pretrained("thenlper/gte-small")
model = AutoModel.from_pretrained("thenlper/gte-small")

# Tokenize the input texts
batch_dict = tokenizer(input_texts, max_length=512, padding=True, truncation=True, return_tensors='pt')

outputs = model(**batch_dict)


config.json: 100%|██████████| 583/583 [00:00<00:00, 2.97MB/s]
/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.11/lib/python3.11/site-packages/transformers/utils/generic.py:309: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
model.safetensors: 100%|██████████| 66.7M/66.7M [00:05<00:00, 11.7MB/s]


In [7]:
outputs.last_hidden_state.shape

torch.Size([4, 10, 384])

In [9]:
from sentence_transformers import SentenceTransformer
from sentence_transformers.util import cos_sim

sentences = ['That is a happy person', 'That is a very happy person']

model = SentenceTransformer('thenlper/gte-large')
embeddings = model.encode(sentences)


.gitattributes: 100%|██████████| 1.52k/1.52k [00:00<00:00, 3.96MB/s]
1_Pooling/config.json: 100%|██████████| 191/191 [00:00<00:00, 521kB/s]
README.md: 100%|██████████| 67.9k/67.9k [00:00<00:00, 361kB/s]
config.json: 100%|██████████| 619/619 [00:00<00:00, 2.50MB/s]
model.safetensors: 100%|██████████| 670M/670M [00:58<00:00, 11.5MB/s] 
onnx/config.json: 100%|██████████| 632/632 [00:00<00:00, 2.50MB/s]
model.onnx: 100%|██████████| 1.34G/1.34G [01:56<00:00, 11.5MB/s]
onnx/special_tokens_map.json: 100%|██████████| 125/125 [00:00<00:00, 539kB/s]
onnx/tokenizer.json: 100%|██████████| 712k/712k [00:00<00:00, 944kB/s]
onnx/tokenizer_config.json: 100%|██████████| 342/342 [00:00<00:00, 1.49MB/s]
onnx/vocab.txt: 100%|██████████| 232k/232k [00:00<00:00, 1.23MB/s]
pytorch_model.bin: 100%|██████████| 670M/670M [00:58<00:00, 11.5MB/s] 
sentence_bert_config.json: 100%|██████████| 57.0/57.0 [00:00<00:00, 177kB/s]
special_tokens_map.json: 100%|██████████| 125/125 [00:00<00:00, 466kB/s]
tokenizer.json: 10

In [11]:
from langchain.vectorstores import FAISS


In [12]:
from langchain.chains.question_answering import load_qa_chain


In [16]:
def train_and_save_doc_space(doc_path = './pdfs', embeddings = None, uploaded_file = False):
    myPdfReader = None

    pdf_files = [os.path.join(doc_path, f) for f in os.listdir(doc_path) if f.endswith('.pdf')]
    st.info(f"Number of pdf files found : {len(pdf_files)}")
    raw_text = ''
    
    if uploaded_file == False:
        for pdf_file in pdf_files:
            myPdfReader = PdfReader(pdf_file)
            for page in myPdfReader.pages:
                raw_text += page.extract_text()
    else:
        myPdfReader = PdfReader(doc_path)
        for page in myPdfReader.pages:
            raw_text += page.extract_text()

    text_splitter = CharacterTextSplitter(
        separator='\n',
        chunk_size=2000,
        chunk_overlap=200,
        length_function = len,
    )
    texts = text_splitter.split_text(raw_text)
    # show the length to the streamlit as log
    st.info(f"Length of the chunks: {len(texts)}")
    st.info("Creating the vector store")
    document_search_space = FAISS.from_texts(texts, embeddings)
    # print("Vector store created")
    st.info("Vector store created")

    document_search_space.save_local('faiss_doc_space')

In [17]:
train_and_save_doc_space(embeddings= model.encode)

2023-12-09 13:39:17.318 
  command:

    streamlit run /home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.11/lib/python3.11/site-packages/ipykernel_launcher.py [ARGUMENTS]


AttributeError: 'function' object has no attribute 'embed_documents'

In [18]:
from langchain.embeddings.huggingface import HuggingFaceEmbeddings



In [19]:
model_name = "sentence-transformers/all-mpnet-base-v2"
model_kwargs = {'device': 'cpu'}
encode_kwargs = {'normalize_embeddings': False}
hf = HuggingFaceEmbeddings(
    model_name=model_name,
    model_kwargs=model_kwargs,
    encode_kwargs=encode_kwargs
)

.gitattributes: 100%|██████████| 1.18k/1.18k [00:00<00:00, 3.57MB/s]
1_Pooling/config.json: 100%|██████████| 190/190 [00:00<00:00, 1.15MB/s]
README.md: 100%|██████████| 10.6k/10.6k [00:00<00:00, 18.1MB/s]
config_sentence_transformers.json: 100%|██████████| 116/116 [00:00<00:00, 377kB/s]
data_config.json: 100%|██████████| 39.3k/39.3k [00:00<00:00, 216kB/s]
pytorch_model.bin: 100%|██████████| 438M/438M [00:38<00:00, 11.4MB/s] 
sentence_bert_config.json: 100%|██████████| 53.0/53.0 [00:00<00:00, 117kB/s]
special_tokens_map.json: 100%|██████████| 239/239 [00:00<00:00, 1.50MB/s]
tokenizer.json: 100%|██████████| 466k/466k [00:00<00:00, 815kB/s]
tokenizer_config.json: 100%|██████████| 363/363 [00:00<00:00, 1.30MB/s]
train_script.py: 100%|██████████| 13.1k/13.1k [00:00<00:00, 23.6MB/s]
vocab.txt: 100%|██████████| 232k/232k [00:00<00:00, 633kB/s]
modules.json: 100%|██████████| 349/349 [00:00<00:00, 1.33MB/s]


In [20]:
train_and_save_doc_space(embeddings= hf)

In [22]:
model_name = "thenlper/gte-large"
model_kwargs = {'device': 'cpu'}
encode_kwargs = {'normalize_embeddings': False}
hf = HuggingFaceEmbeddings(
    model_name=model_name,
    model_kwargs=model_kwargs,
    encode_kwargs=encode_kwargs
)

In [23]:
train_and_save_doc_space(embeddings= hf)

In [24]:
from langchain.chains.question_answering import load_qa_chain


In [25]:
chain1 = load_qa_chain(llm = model)

ValidationError: 2 validation errors for LLMChain
llm
  instance of Runnable expected (type=type_error.arbitrary_type; expected_arbitrary_type=Runnable)
llm
  instance of Runnable expected (type=type_error.arbitrary_type; expected_arbitrary_type=Runnable)

In [40]:
from langchain.llms.huggingface_endpoint import HuggingFaceEndpoint
from langchain.llms.huggingface_hub import HuggingFaceHub

In [27]:
from langchain import llms

In [29]:
# list all the models 
dir(llms)

['Any',
 'BaseLLM',
 'Callable',
 'Dict',
 'Type',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__getattr__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '_import_ai21',
 '_import_aleph_alpha',
 '_import_amazon_api_gateway',
 '_import_anthropic',
 '_import_anyscale',
 '_import_arcee',
 '_import_aviary',
 '_import_azure_openai',
 '_import_azureml_endpoint',
 '_import_baidu_qianfan_endpoint',
 '_import_bananadev',
 '_import_baseten',
 '_import_beam',
 '_import_bedrock',
 '_import_bittensor',
 '_import_cerebriumai',
 '_import_chatglm',
 '_import_clarifai',
 '_import_cohere',
 '_import_ctransformers',
 '_import_ctranslate2',
 '_import_databricks',
 '_import_databricks_chat',
 '_import_deepinfra',
 '_import_deepsparse',
 '_import_edenai',
 '_import_fake',
 '_import_fireworks',
 '_import_forefrontai',
 '_import_gigachat',
 '_import_google_palm',
 '_import_gooseai',
 '_import_gpt4all',
 '_import_gradient_ai',
 '_import_huggingface_endpoint'

In [34]:
dir(llms.huggingface_hub.LLM)

['Config',
 'InputType',
 'OutputType',
 '__abstractmethods__',
 '__annotations__',
 '__call__',
 '__class__',
 '__class_getitem__',
 '__class_vars__',
 '__config__',
 '__custom_root_type__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__exclude_fields__',
 '__fields__',
 '__fields_set__',
 '__format__',
 '__ge__',
 '__get_validators__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__include_fields__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__json_encoder__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__or__',
 '__orig_bases__',
 '__parameters__',
 '__post_root_validators__',
 '__pre_root_validators__',
 '__pretty__',
 '__private_attributes__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__repr_args__',
 '__repr_name__',
 '__repr_str__',
 '__rich_repr__',
 '__ror__',
 '__schema_cache__',
 '__setattr__',
 '__setstate__',
 '__signature__',
 '__sizeof__',
 '__slots__',
 '__str__',
 '__subclasshook__',
 '__try_updat

In [39]:
model_name = "thenlper/gte-large"
hf = HuggingFaceEndpoint(
    model_name=model_name,
)

ValidationError: 2 validation errors for HuggingFaceEndpoint
model_name
  extra fields not permitted (type=value_error.extra)
__root__
  Did not find huggingfacehub_api_token, please add an environment variable `HUGGINGFACEHUB_API_TOKEN` which contains it, or pass  `huggingfacehub_api_token` as a named parameter. (type=value_error)

In [43]:
HuggingFaceHub(model_name=model_name)

ValidationError: 2 validation errors for HuggingFaceHub
model_name
  extra fields not permitted (type=value_error.extra)
__root__
  Did not find huggingfacehub_api_token, please add an environment variable `HUGGINGFACEHUB_API_TOKEN` which contains it, or pass  `huggingfacehub_api_token` as a named parameter. (type=value_error)

In [2]:
from langchain.document_loaders import OnlinePDFLoader
loader = OnlinePDFLoader("https://www.dropbox.com/scl/fo/ldsupnijn2a8nh7modnnj/h?rlkey=8zt0fnl8yeco3lnth9nzimcwm&dl=0")



In [11]:
loader.load()

/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
No /Root object! - Is this really a PDF?
Traceback (most recent call last):
  File "/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/unstructured/partition/pdf.py", line 255, in partition_pdf_or_image
    extracted_elements = extractable_elements(
  File "/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/unstructured/partition/pdf.py", line 200, in extractable_elements
    return _partition_pdf_with_pdfminer(
  File "/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/unstructured/utils.py", line 214, in wrapper
    return func(*args, **kwargs)
  File "/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/unstructured/partition/pdf.

PDFPageCountError: Unable to get page count.
Syntax Warning: May not be a PDF file (continuing anyway)
Syntax Error (2): Illegal character <21> in hex string
Syntax Error (4): Illegal character <4f> in hex string
Syntax Error (6): Illegal character <54> in hex string
Syntax Error (7): Illegal character <59> in hex string
Syntax Error (8): Illegal character <50> in hex string
Syntax Error (11): Illegal character <68> in hex string
Syntax Error (12): Illegal character <74> in hex string
Syntax Error (13): Illegal character <6d> in hex string
Syntax Error (14): Illegal character <6c> in hex string
Syntax Error (18): Illegal character <68> in hex string
Syntax Error (19): Illegal character <74> in hex string
Syntax Error (20): Illegal character <6d> in hex string
Syntax Error (21): Illegal character <6c> in hex string
Syntax Error (24): Illegal character <6c> in hex string
Syntax Error (26): Illegal character <73> in hex string
Syntax Error (27): Illegal character <73> in hex string
Syntax Error (28): Illegal character <3d> in hex string
Syntax Error (29): Illegal character <22> in hex string
Syntax Error (30): Illegal character <6d> in hex string
Syntax Error (33): Illegal character <73> in hex string
Syntax Error (34): Illegal character <74> in hex string
Syntax Error (35): Illegal character <72> in hex string
Syntax Error (36): Illegal character <6f> in hex string
Syntax Error (38): Illegal character <67> in hex string
Syntax Error (39): Illegal character <6c> in hex string
Syntax Error (40): Illegal character <6f> in hex string
Syntax Error (43): Illegal character <6c> in hex string
Syntax Error (44): Illegal character <2d> in hex string
Syntax Error (45): Illegal character <68> in hex string
Syntax Error (50): Illegal character <72> in hex string
Syntax Error (51): Illegal character <22> in hex string
Syntax Error (53): Illegal character <78> in hex string
Syntax Error (54): Illegal character <6d> in hex string
Syntax Error (55): Illegal character <6c> in hex string
Syntax Error (56): Illegal character <6e> in hex string
Syntax Error (57): Illegal character <73> in hex string
Syntax Error (58): Illegal character <3d> in hex string
Syntax Error (59): Illegal character <22> in hex string
Syntax Error (60): Illegal character <68> in hex string
Syntax Error (61): Illegal character <74> in hex string
Syntax Error (62): Illegal character <74> in hex string
Syntax Error (63): Illegal character <70> in hex string
Syntax Error (64): Illegal character <3a> in hex string
Syntax Error (65): Illegal character <2f> in hex string
Syntax Error (66): Illegal character <2f> in hex string
Syntax Error (67): Illegal character <77> in hex string
Syntax Error (68): Illegal character <77> in hex string
Syntax Error (69): Illegal character <77> in hex string
Syntax Error (70): Illegal character <2e> in hex string
Syntax Error (71): Illegal character <77> in hex string
Syntax Error (73): Illegal character <2e> in hex string
Syntax Error (74): Illegal character <6f> in hex string
Syntax Error (75): Illegal character <72> in hex string
Syntax Error (76): Illegal character <67> in hex string
Syntax Error (77): Illegal character <2f> in hex string
Syntax Error (82): Illegal character <2f> in hex string
Syntax Error (83): Illegal character <78> in hex string
Syntax Error (84): Illegal character <68> in hex string
Syntax Error (85): Illegal character <74> in hex string
Syntax Error (86): Illegal character <6d> in hex string
Syntax Error (87): Illegal character <6c> in hex string
Syntax Error (88): Illegal character <22> in hex string
Syntax Error (90): Illegal character <6c> in hex string
Syntax Error (92): Illegal character <6e> in hex string
Syntax Error (93): Illegal character <67> in hex string
Syntax Error (94): Illegal character <3d> in hex string
Syntax Error (95): Illegal character <22> in hex string
Syntax Error (97): Illegal character <6e> in hex string
Syntax Error (98): Illegal character <22> in hex string
Syntax Error (102): Illegal character <68> in hex string
Syntax Error (108): Illegal character <6d> in hex string
Syntax Error (110): Illegal character <74> in hex string
Syntax Error (114): Illegal character <68> in hex string
Syntax Error (116): Illegal character <72> in hex string
Syntax Error (117): Illegal character <73> in hex string
Syntax Error (119): Illegal character <74> in hex string
Syntax Error (120): Illegal character <3d> in hex string
Syntax Error (121): Illegal character <22> in hex string
Syntax Error (122): Illegal character <75> in hex string
Syntax Error (123): Illegal character <74> in hex string
Syntax Error (125): Illegal character <2d> in hex string
Syntax Error (127): Illegal character <22> in hex string
Syntax Error (129): Illegal character <2f> in hex string
Syntax Error (133): Illegal character <6d> in hex string
Syntax Error (135): Illegal character <74> in hex string
Syntax Error (139): Illegal character <6f> in hex string
Syntax Error (140): Illegal character <6e> in hex string
Syntax Error (141): Illegal character <74> in hex string
Syntax Error (143): Illegal character <6e> in hex string
Syntax Error (144): Illegal character <74> in hex string
Syntax Error (145): Illegal character <3d> in hex string
Syntax Error (146): Illegal character <22> in hex string
Syntax Error (147): Illegal character <6e> in hex string
Syntax Error (148): Illegal character <6f> in hex string
Syntax Error (149): Illegal character <69> in hex string
Syntax Error (150): Illegal character <6e> in hex string
Syntax Error (153): Illegal character <78> in hex string
Syntax Error (154): Illegal character <2c> in hex string
Syntax Error (156): Illegal character <6e> in hex string
Syntax Error (157): Illegal character <6f> in hex string
Syntax Error (159): Illegal character <6f> in hex string
Syntax Error (160): Illegal character <6c> in hex string
Syntax Error (161): Illegal character <6c> in hex string
Syntax Error (162): Illegal character <6f> in hex string
Syntax Error (163): Illegal character <77> in hex string
Syntax Error (164): Illegal character <2c> in hex string
Syntax Error (166): Illegal character <6e> in hex string
Syntax Error (167): Illegal character <6f> in hex string
Syntax Error (168): Illegal character <69> in hex string
Syntax Error (169): Illegal character <6d> in hex string
Syntax Error (171): Illegal character <67> in hex string
Syntax Error (173): Illegal character <69> in hex string
Syntax Error (174): Illegal character <6e> in hex string
Syntax Error (177): Illegal character <78> in hex string
Syntax Error (178): Illegal character <22> in hex string
Syntax Error (180): Illegal character <6e> in hex string
Syntax Error (182): Illegal character <6d> in hex string
Syntax Error (184): Illegal character <3d> in hex string
Syntax Error (185): Illegal character <22> in hex string
Syntax Error (186): Illegal character <72> in hex string
Syntax Error (187): Illegal character <6f> in hex string
Syntax Error (189): Illegal character <6f> in hex string
Syntax Error (190): Illegal character <74> in hex string
Syntax Error (191): Illegal character <73> in hex string
Syntax Error (192): Illegal character <22> in hex string
Syntax Error (194): Illegal character <2f> in hex string
Syntax Error (198): Illegal character <6d> in hex string
Syntax Error (200): Illegal character <74> in hex string
Syntax Error (204): Illegal character <6f> in hex string
Syntax Error (205): Illegal character <6e> in hex string
Syntax Error (206): Illegal character <74> in hex string
Syntax Error (208): Illegal character <6e> in hex string
Syntax Error (209): Illegal character <74> in hex string
Syntax Error (210): Illegal character <3d> in hex string
Syntax Error (211): Illegal character <22> in hex string
Syntax Error (212): Illegal character <77> in hex string
Syntax Error (213): Illegal character <69> in hex string
Syntax Error (215): Illegal character <74> in hex string
Syntax Error (216): Illegal character <68> in hex string
Syntax Error (217): Illegal character <3d> in hex string
Syntax Error (220): Illegal character <76> in hex string
Syntax Error (221): Illegal character <69> in hex string
Syntax Error (224): Illegal character <2d> in hex string
Syntax Error (225): Illegal character <77> in hex string
Syntax Error (226): Illegal character <69> in hex string
Syntax Error (228): Illegal character <74> in hex string
Syntax Error (229): Illegal character <68> in hex string
Syntax Error (230): Illegal character <2c> in hex string
Syntax Error (232): Illegal character <69> in hex string
Syntax Error (233): Illegal character <6e> in hex string
Syntax Error (234): Illegal character <69> in hex string
Syntax Error (235): Illegal character <74> in hex string
Syntax Error (236): Illegal character <69> in hex string
Syntax Error (238): Illegal character <6c> in hex string
Syntax Error (239): Illegal character <2d> in hex string
Syntax Error (240): Illegal character <73> in hex string
Syntax Error (243): Illegal character <6c> in hex string
Syntax Error (245): Illegal character <3d> in hex string
Syntax Error (247): Illegal character <22> in hex string
Syntax Error (249): Illegal character <6e> in hex string
Syntax Error (251): Illegal character <6d> in hex string
Syntax Error (253): Illegal character <3d> in hex string
Syntax Error (254): Illegal character <22> in hex string
Syntax Error (255): Illegal character <76> in hex string
Syntax Error (256): Illegal character <69> in hex string
Syntax Error (258): Illegal character <77> in hex string
Syntax Error (259): Illegal character <70> in hex string
Syntax Error (260): Illegal character <6f> in hex string
Syntax Error (261): Illegal character <72> in hex string
Syntax Error (262): Illegal character <74> in hex string
Syntax Error (263): Illegal character <22> in hex string
Syntax Error (265): Illegal character <2f> in hex string
Syntax Error: Couldn't find trailer dictionary
Syntax Error: Couldn't find trailer dictionary
Syntax Error: Couldn't read xref table


In [10]:
!pip install unstructured_inference

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.2/19.2 MB 8.8 MB/s eta 0:00:00:00:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.7/45.7 kB 634.0 kB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.0/46.0 kB 1.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.2/42.2 kB 395.6 kB/s eta 0:00:00a 0:00:01
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.8/86.8 kB 11.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 987.2 kB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.5/79.5 kB 1.3 MB/s eta 0:00:00ta 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.6/45.6 kB 955.1 kB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 117.0/117.0 kB 2.5 MB/s eta 0:00:00a 0:00:01
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 156.2/156.2 kB 1.6 MB/s eta 0:00:0000:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [8]:
!pip install unstructured_pytesseract

In [6]:
!pip install opencv-python

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.7/61.7 MB 8.5 MB/s eta 0:00:00:00:0100:01


In [ ]:
pip install opencv-python

In [4]:
!pip install unstructured
!pip install pdf2image
!pip install pikepdf
!pip install pypdf
!pip install pdfminer.six

  Using cached unstructured-0.11.2-py3-none-any.whl.metadata (25 kB)
  Using cached chardet-5.2.0-py3-none-any.whl.metadata (3.4 kB)
  Using cached filetype-1.2.0-py2.py3-none-any.whl (19 kB)
  Using cached python_magic-0.4.27-py2.py3-none-any.whl (13 kB)
  Using cached tabulate-0.9.0-py3-none-any.whl (35 kB)
  Using cached beautifulsoup4-4.12.2-py3-none-any.whl (142 kB)
  Using cached emoji-2.9.0-py2.py3-none-any.whl.metadata (5.3 kB)
  Using cached python_iso639-2023.6.15-py3-none-any.whl.metadata (13 kB)
  Using cached langdetect-1.0.9.tar.gz (981 kB)
  Preparing metadata (setup.py) ... done
  Using cached backoff-2.2.1-py3-none-any.whl (15 kB)
  Using cached soupsieve-2.5-py3-none-any.whl.metadata (4.7 kB)
Using cached unstructured-0.11.2-py3-none-any.whl (1.7 MB)
Using cached chardet-5.2.0-py3-none-any.whl (199 kB)
Using cached emoji-2.9.0-py2.py3-none-any.whl (397 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.0/8.0 MB 9.1 MB/s eta 0:00:0000:0100:01
Using cached python_iso639-

# Huggingface model Try -1


In [1]:
import os
os.environ["HUGGINGFACEHUB_API_TOKEN"] = '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'
# from langchain.llms import OpenAI
from langchain.llms import HuggingFaceHub
from sentence_transformers import SentenceTransformer
from sentence_transformers.util import cos_sim

from langchain.prompts import (PromptTemplate, FewShotPromptTemplate)
from langchain.memory import ConversationBufferMemory 
from langchain.chains import (LLMChain, SimpleSequentialChain, SequentialChain)

# from langchain.embeddings.openai import OpenAIEmbeddings
from langchain.embeddings.huggingface import HuggingFaceEmbeddings

from langchain.text_splitter import CharacterTextSplitter
from langchain.vectorstores import FAISS
from langchain.chains.question_answering import load_qa_chain
from typing_extensions import Concatenate
from PyPDF2 import PdfReader
from langchain.document_loaders import OnlinePDFLoader



import streamlit as st
import pandas as pd

/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


#### Trainning and loading the doc space

In [2]:
def train_and_save_doc_space(doc_path = './pdfs', embeddings = None, uploaded_file = False, name='faiss_doc_space'):
    myPdfReader = None

    pdf_files = [os.path.join(doc_path, f) for f in os.listdir(doc_path) if f.endswith('.pdf')]
    st.info(f"Number of pdf files found : {len(pdf_files)}")
    raw_text = ''
    
    if uploaded_file == False:
        for pdf_file in pdf_files:
            myPdfReader = PdfReader(pdf_file)
            for page in myPdfReader.pages:
                raw_text += page.extract_text()
    else:
        myPdfReader = PdfReader(doc_path)
        for page in myPdfReader.pages:
            raw_text += page.extract_text()

    text_splitter = CharacterTextSplitter(
        separator='\n',
        chunk_size=512,
        chunk_overlap=150,
        length_function = len,
    )
    texts = text_splitter.split_text(raw_text)
    # show the length to the streamlit as log
    st.info(f"Length of the chunks: {len(texts)}")
    st.info("Creating the vector store")
    document_search_space = FAISS.from_texts(texts, embeddings)
    # print("Vector store created")
    st.info("Vector store created")

    document_search_space.save_local(name)

def load_doc_space(name = 'faiss_doc_space', embeddings = None):
    return FAISS.load_local(name, embeddings)


#### Embeddings

1. all-MiniLM-L6-v2
This is a sentence-transformers model: It maps sentences & paragraphs to a 384 dimensional dense vector space and can be used for tasks like clustering or semantic search.

2. jina-embeddings-v2-base-en is an English, monolingual embedding model supporting 8192 sequence length. It is based on a Bert architecture (JinaBert) that supports the symmetric bidirectional variant of ALiBi to allow longer sequence length. The backbone jina-bert-v2-base-en is pretrained on the C4 dataset. The model is further trained on Jina AI's collection of more than 400 millions of sentence pairs and hard negatives. These pairs were obtained from various domains and were carefully selected through a thorough cleaning process. (leaderboard MTEB - 24, 0.27 GB)

3. jina-embeddings-v2-base-en: The Small model comprises 33 million parameters, whereas the Base model comprises 137 million parameters. (leaderboard MTEB - 39/ 0.07GB)

4. thenlper/gte-large: 


In [3]:
model_name = "thenlper/gte-large"
model_kwargs = {'device': 'cpu'}
encode_kwargs = {'normalize_embeddings': False}

gte_embedding = HuggingFaceEmbeddings(
    model_name=model_name,
    model_kwargs=model_kwargs,
    encode_kwargs=encode_kwargs
)

mini_lm_l6_embedding =HuggingFaceEmbeddings(model_name='sentence-transformers/all-MiniLM-L6-v2')



In [4]:
jina_embedding = HuggingFaceEmbeddings(model_name='jinaai/jina-embeddings-v2-small-en', )

In [5]:
embedding2 = HuggingFaceEmbeddings(model_name='sentence-transformers/all-mpnet-base-v2')

In [6]:
train_and_save_doc_space(doc_path="./pdfs/", embeddings=gte_embedding, name='faiss_doc_space_gte')
train_and_save_doc_space(doc_path="./pdfs/", embeddings=mini_lm_l6_embedding, name='faiss_doc_space_mini_lm_l6')
train_and_save_doc_space(doc_path="./pdfs/", embeddings=jina_embedding, name='faiss_doc_space_jina')
train_and_save_doc_space(doc_path="./pdfs/", embeddings=embedding2, name='faiss_doc_space_mpnet_base_v2')
document_search_space1 = load_doc_space(embeddings=gte_embedding, name = 'faiss_doc_space_gte')
document_search_space2 = load_doc_space(embeddings=mini_lm_l6_embedding, name = 'faiss_doc_space_mini_lm_l6')
document_search_space3 = load_doc_space(embeddings=jina_embedding, name = 'faiss_doc_space_jina')    
document_search_space4 = load_doc_space(embeddings=embedding2, name = 'faiss_doc_space_mpnet_base_v2')


2023-12-11 01:10:48.662 
  command:

    streamlit run /home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/ipykernel_launcher.py [ARGUMENTS]


# Document space training

In [15]:
query = "What did lancho write in the second letter to god?"
docs1 = document_search_space1.similarity_search(query,k = 5)
docs2 = document_search_space2.similarity_search(query,k = 5)
docs3 = document_search_space3.similarity_search(query,k = 5)
docs4 = document_search_space4.similarity_search(query,k = 5)
actual_ans = 'he made a request to god for sending the remaining 30 pesos'

#### Models

In [7]:

llm_google_flan_t5=HuggingFaceHub(
        repo_id="google/flan-t5-xxl", 
        model_kwargs={"temperature":0.2, 
                    #   "max_length":512
                    }
        )

chain2 = load_qa_chain( llm=llm_google_flan_t5, chain_type="stuff")

llm_mistral_7b = HuggingFaceHub(
        repo_id="mistralai/Mistral-7B-v0.1",
        model_kwargs={"temperature":0.5, 
                        #   "max_length":512
                        }
        )

chain1 = load_qa_chain( llm=llm_mistral_7b, chain_type="stuff")

/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


In [13]:
llm_mistral_7b_instruct = HuggingFaceHub(
        repo_id="mistralai/Mistral-7B-Instruct-v0.1",
        model_kwargs={"temperature":0.5, 
                        }
        )
chain3 = load_qa_chain( llm=llm_mistral_7b_instruct, chain_type="stuff")


/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


In [34]:
llm_longformer_base_4k = HuggingFaceHub(
        repo_id="Udit191/autotrain-summarization_bart_longformer-54164127153",
        model_kwargs={"temperature":0.5, 
                        }
        )
chain4 = load_qa_chain( llm=llm_longformer_base_4k, chain_type="stuff")
chain4.run(input_documents = docs1, question = query)

/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


KeyboardInterrupt: 

In [35]:
#meta-llama/Llama-2-7b-chat-hf
llm_llama_2_7b_chat_hf = HuggingFaceHub(
        repo_id="meta-llama/Llama-2-7b-chat-hf",
        model_kwargs={"temperature":0.5, 
                        }
        )
chain5 = load_qa_chain( llm=llm_llama_2_7b_chat_hf, chain_type="stuff")
print("chain5 loaded")
chain5.run(input_documents = docs1, question = query)

/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


GatedRepoError: 403 Client Error. (Request ID: Root=1-65762cf4-32829033145a4bd813077a8a;b0d0e79f-4e82-4868-98db-dd8c32173e3c)

Cannot access gated repo for url https://huggingface.co/api/models/meta-llama/Llama-2-7b-chat-hf.
Your request to access model meta-llama/Llama-2-7b-chat-hf is awaiting a review from the repo authors.

In [ ]:
# meta-llama/Llama-2-7b-chat
llm_llama_2_7b_chat = HuggingFaceHub(
        repo_id="meta-llama/Llama-2-7b-chat",
        model_kwargs={"temperature":0.5, 
                        }
        )
chain6 = load_qa_chain( llm=llm_llama_2_7b_chat, chain_type="stuff")
print("chain6 loaded")
chain6.run(input_documents = docs1, question = query)

/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


GatedRepoError: 403 Client Error. (Request ID: Root=1-65761856-59bd29a51d9bba06047d23d4;c60cae52-3f57-49cb-a6b1-dbefc29627da)

Cannot access gated repo for url https://huggingface.co/api/models/meta-llama/Llama-2-7b-chat.
Your request to access model meta-llama/Llama-2-7b-chat is awaiting a review from the repo authors.

In [36]:
#Intel/neural-chat-7b-v3-1
llm_neural_chat_7b_v3_1 = HuggingFaceHub(
        repo_id="Intel/neural-chat-7b-v3-3",
        model_kwargs={"temperature":0.5, 
                      
                        }
        )
chain7 = load_qa_chain( llm=llm_neural_chat_7b_v3_1, chain_type="stuff")
chain7.run(input_documents = docs1, question = query)

/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


KeyboardInterrupt: 

In [30]:
# microsoft/Orca-2-13b
llm_orca_2_13b = HuggingFaceHub(
        repo_id="microsoft/Orca-2-13b",
        model_kwargs={"temperature":0.5, 
                        }
        )
chain8 = load_qa_chain( llm=llm_orca_2_13b, chain_type="stuff")
print("chain8 loaded")
chain8.run(input_documents = docs1, question = query)

/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


chain8 loaded


ValueError: Error raised by inference API: Model microsoft/Orca-2-13b time out

In [32]:
# google/mt5-base
llm_mt5_base = HuggingFaceHub(
        repo_id="google/mt5-base",
        model_kwargs={"temperature":0.5, 
                        }
        )
chain9 = load_qa_chain( llm=llm_mt5_base, chain_type="stuff")
print("chain9 loaded")
print(chain9.run(input_documents = docs1, question = query))

/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


chain9 loaded
<extra_id_0>. . . . .


In [33]:
# google/umt5-small
llm_umt5_small = HuggingFaceHub(
        repo_id="google/umt5-small",
        model_kwargs={"temperature":0.5, 
                        }
        )
chain10 = load_qa_chain( llm=llm_umt5_small, chain_type="stuff")
print("chain10 loaded")
print(chain10.run(input_documents = docs1, question = query))

/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


chain10 loaded
, he wrote,., he wrote,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,


In [37]:
# bigscience/bloomz-560m
llm_bloomz_560m = HuggingFaceHub(
        repo_id="bigscience/bloomz-560m",
        model_kwargs={"temperature":0.5, 
                      
                        }
        )
chain11 = load_qa_chain( llm=llm_bloomz_560m, chain_type="stuff")
chain11.run(input_documents = docs1, question = query)

/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


' God'

In [38]:
# google/long-t5-tglobal-large
llm_long_t5_tglobal_large = HuggingFaceHub(
        repo_id="google/long-t5-tglobal-large",
        model_kwargs={"temperature":0.5, 
                      
                        }
        )
chain12 = load_qa_chain( llm=llm_long_t5_tglobal_large, chain_type="stuff")
chain12.run(input_documents = docs1, question = query)

/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


'1256%ensembleensembleräumt population populationstairAVE lenderlitis It lender lender Two Interest Interest Interest Interest'

In [39]:
# google/long-t5-tglobal-base
llm_long_t5_tglobal_base = HuggingFaceHub(
        repo_id="google/long-t5-tglobal-base",
        model_kwargs={"temperature":0.5, 
                      
                        }
        )
chain13 = load_qa_chain( llm=llm_long_t5_tglobal_base, chain_type="stuff")
chain13.run(input_documents = docs1, question = query)


/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


'When he finished, he went to the window to buy a stamp which he'

In [40]:
# google/long-t5-tglobal-xl
llm_long_t5_tglobal_xl = HuggingFaceHub(
        repo_id="google/long-t5-tglobal-xl",
        model_kwargs={"temperature":0.5, 
                      
                        }
        )
chain14 = load_qa_chain( llm=llm_long_t5_tglobal_xl, chain_type="stuff")
chain14.run(input_documents = docs1, question = query)


/home/dosisiddhesh/LANGCHAIN_EXP/env_lc_3.8/lib/python3.8/site-packages/huggingface_hub/utils/_deprecation.py:127: FutureWarning: '__init__' (from 'huggingface_hub.inference_api') is deprecated and will be removed from version '1.0'. `InferenceApi` client is deprecated in favor of the more feature-complete `InferenceClient`. Check out this guide to learn how to convert your script to use it: https://huggingface.co/docs/huggingface_hub/guides/inference#legacy-inferenceapi-client.
  warnings.warn(warning_message, FutureWarning)


KeyboardInterrupt: 

In [25]:
chain4.run(input_documents = docs1, question = query)

KeyboardInterrupt: 

In [16]:
chain1.run(input_documents = docs1, question = query)

ValueError: Error raised by inference API: Internal Server Error

In [14]:
chain3.run(input_documents = docs1, question = query)

' Lancho wrote "Don\'t be so upset, even though this seems like a total loss. Remember , no one dies of hunger ." in the second letter to God.'

In [29]:
print(chain2.run(input_documents = docs1, question = query))

I need a hundred pesos in order to sow my field again and to live until the crop comes, because the hailstorm....”


In [30]:
print(chain2.run(input_documents = docs2, question = query))

“God,” he wrote, “if you don’t help me, my family and I will go hungry this year. I need a hundr ed pesos in order to sow my field again and to live until the crop comes, because the hailstorm....”


In [31]:
print(chain2.run(input_documents = docs3, question = query))

“God,” he wrote, “if you don’t help me, my family and I will go hungry this year. I need a hundr ed pesos in order to sow my field again and to live until the crop comes, because the hailstorm....”


In [35]:
print(chain2.run(input_documents = docs4, question = query))

Don’t be so upset, even though this seems like a total loss. Remember, no one dies of hunger.” “That’s what they say: no one dies of hunger.” All through the night, Lencho thought only of his one hope: the help of God, whose eyes, as he had been instructed, see everything, even what is deep in one’s conscience.


In [33]:
print(chain1.run(input_documents = docs1, question = query))

ValueError: Error raised by inference API: Internal Server Error